### DISCUSSION TITLE
Query to check if Document ID exists

### DISCUSSION THREAD

**Discussion by JediThumper** [SchneiderInfosystems/FB4D#171] [Q&A] [answered]
Hi,

Hope you had a great Christmas.

I want to check if a certain document exists.

My Query is:

 Query := TStructuredQuery.CreateForCollection('Practitioners')
    .QueryForFieldFilter(TQueryFilter.StringFieldFilter('Document ID', TWhereOperator.woEqual, APractitionerID));
DocsReturned := fDataBase.RunQuerySynchronous(Query);

Result := (DocsRecturned.Count =1);

But I get an error:

Project Denneroll.apk raised exception class EFirebaseResponse with message 'Invalid property path "Document ID". Unquoted property paths must match regex ([a-zA-Z_][a-zA-Z_0-9]*), and quoted property paths must match regex (`(?:[^`\\]|(?:\\.))+`)'.

I know I can save the Document ID in the Document and query that. Which works. But the query builder in Firebase allows a query on Document ID so there must be a way.

As always appreciate your advice.

Happy New Year

Alan

**Comment by SchneiderInfosystems** [comment]
Hi Alain
Can you confirm that the reason is the space character between "Document" and "ID"? If this is the case, I have to check if the parameter need to be set in quotes inside `TQueryFilter.StringFieldFilter`.
br
Christoph

**Reply by JediThumper** [reply]
Hi Christoph,

Thanks for the speedy response. 

I can confirm that without the space it does not generate an error.

However, If you look at the screenshot Firebase automatically names this column  "Document ID" with a space. 

I can not change the name.
 
![Screenshot 2024-01-08 172725](https://github.com/SchneiderInfosystems/FB4D/assets/71242335/5ece3b61-149b-4c3c-8f06-8924ea448ad3)

**Comment by SchneiderInfosystems** [comment]
Can you shortly check, if the following approach is working?

Change the line 1413 of Firestore.pas:
```
class function TQueryFilter.StringFieldFilter(const WhereFieldPath: string;
  WhereOperator: TWhereOperator; const WhereValue: string): IQueryFilter;
begin
  result := TQueryFilter.Create(WhereFieldPath, WhereValue, WhereOperator);
  result.AddPair('field',
    TJSONObject.Create(TJSONPair.Create('fieldPath', WhereFieldPath)));
```
to 
```
  result.AddPair('field',
    TJSONObject.Create(TJSONPair.Create('fieldPath', '"' + WhereFieldPath + '"')));
```

**Reply by JediThumper** [reply]
I applied the change but no difference - same error.

```
class function TQueryFilter.StringFieldFilter(const WhereFieldPath: string;
  WhereOperator: TWhereOperator; const WhereValue: string): IQueryFilter;
begin
  result := TQueryFilter.Create(WhereFieldPath, WhereValue, WhereOperator);
  result.AddPair('field',
    TJSONObject.Create(TJSONPair.Create('fieldPath', '"' + WhereFieldPath + '"')));
  result.AddPair('op', FILTER_OERATION[WhereOperator]);
  result.AddPair('value',
    TJSONObject.Create(TJSONPair.Create('stringValue', WhereValue)));
end;
```

**Reply by JediThumper** [reply]
Should I query on DocumentName?

Document.DocumentName(False) is the field I want to query.

**Reply by SchneiderInfosystems** [reply, marked as answer]
Are you searching a distinct document defined by its path and document ID? 
Why not take the simple approach via `TFirestoreDatabase.Get(['Practitioners', APractitionerID]`?

**Reply by JediThumper** [reply]
Thank you. 
This works. I am so stuck in SQL mode that I think only in Queries. I never considered a simple Get.
Sorry you had to put the effort into changing code.
Alan

**Comment by SchneiderInfosystems** [comment]
And this?
```
  result.AddPair('field',
    TJSONObject.Create(TJSONPair.Create('fieldPath', '&quot;' + WhereFieldPath + '&quot;')));
```

**Reply by JediThumper** [reply]
Sorry Doesn't work:

Project iOsDenneroll.apk raised exception class EFirebaseResponse with message 'Invalid property path "&quot;Document ID&quot;". Unquoted property paths must match regex ([a-zA-Z_][a-zA-Z_0-9]*), and quoted property paths must match regex (`(?:[^`\\]|(?:\\.))+`)'.

Also breaks for field with no space in name

Project iOsDenneroll.apk raised exception class EFirebaseResponse with message 'Invalid property path "&quot;PractitionerID&quot;". Unquoted property paths must match regex ([a-zA-Z_][a-zA-Z_0-9]*), and quoted property paths must match regex (`(?:[^`\\]|(?:\\.))+`)'.

**Reply by SchneiderInfosystems** [reply]
Can you show the document and its field in the Firebase Console?

**Comment by SchneiderInfosystems** [comment]
Thank you Alain for the feedback. I wish you much success with this library!
On the otherside, remember that this project cannot be sustained without sponsors. So if you are building a commercial product, you should donate for further support like this discussion. If you have a non-commercial project, tell us about it so I can continue to support you well. And by giving a star you help this project for more visibility.

**Reply by JediThumper** [reply]
Hi Christoph,
At the moment this is a hobby project, that will be a free app.
FYI - I'm a Chiropractor and this will be an app to track and encourage patients to keep with a specific rehab exercise.
I am hoping to monetize it in the future (by providing to other practitioners )at which time I will definitely contribute.

**Reply by SchneiderInfosystems** [reply]
Hi Alain,
All the best and if you publish, please drop me a line. In the longer term, I would like to create a directory of apps based on Fb4D. This could also help your app to increase its visibility. 
Regards Christoph
